packages/backend/src/apps/m365-excel/actions/find-first-table-row/index.ts

```diff
@@ -0,0 +1,247 @@
+      variables: true,
+    },
+  ],
+
+  getDataOutMetadata,
+
+  async run($) {
+    const parametersParseResult = parametersSchema.safeParse($.step.parameters)
+
+    if (parametersParseResult.success === false) {
+      throw new StepError(
+        'There was a problem with the input.',
+        parametersParseResult.error.issues[0].message,
+        $.step?.position,
+        $.app.name,
+      )
+    }
+
+    /**
+     * ** NOTE **
+     *
+     * This does a _full table scan_ in order to minimize the number of queries
+     * made to M365. This decision was made because a full table scan more than
+     * halves the queries needed (from 5 to 2), and we are more constrained by
+     * our _very very low_ M365 rate limit than our available compute.
+     *
+     * A comparison of queries needed:
+     *
+     * |      WITHOUT FULL TABLE SCAN     |        WITH FULL TABLE SCAN        |
+     * | -------------------------------- | ---------------------------------- |
+     * | 1. Create workbook session.      | 1. Create workbook session         |
+     * | 2. Query table address range.    | 2. Query all table rows starting   |
+     * | 3. Perform a MATCH() formula     |    from header row, up to the      |
+     * |    call on table's address       |    MAX_ROW-th row. Then compute    |
+     * |    range to find the 1st row     |    the action output using this    |
+     * |    number which satisfies the    |    alone.                          |
+     * |    search criteria.              |                                    |
+     * | 4. Query table row data using    |                                    |
+     * |    row number from MATCH's       |                                    |
+     * |    output                        |                                    |
+     * | 5. Query table colum names. Use  |                                    |
+     * |    that and the output from step |                                    |
+     * |    step 4 to generate the action |                                    |
+     * |    output                        |                                    |
+     *
+     * The main pitfalls and mitigations for this are:
+     *   1. Although we only fetch up to MAX_ROWS rows, fetching large tables
+     *      with many columns could take a lot of time. This should be very
+     *      rare, but if this ever becomes common we could mitigate it by
+     *      querying only the searched-for column first, then making another
+     *      query to get the full row data for the matched row.
+     *   2. Long strings in cells may cause our scan to be slow. This should
+     *      also be rare; if this ever becomes an issue, we could restrict the
+     *      max length of the value to match.
+     **/
+
+    const { fileId, tableId, columnName, valueToFind } =
+      parametersParseResult.data
+
+    const session = await WorkbookSession.acquire($, fileId)
+    const tableRows = tableRowResponseSchema.parse(
```

Comment:
Tested on a file with 10k rows and 100 columns; zod parsing perf hit was acceptable (~600ms for zod vs ~500ms compared to directly casting via generics; only ~100ms more).